### PR TITLE
Feature/better remove algo

### PR DESCRIPTION
**Pull request by Asergey91**
this new deletion algorithm will basically delete all nodes that it discovers opening tags for in between the 2 range positions, and then migrate the siblings from the end of the range up to the begining of the range
(square brackets denote the range)
from this
```
      <modelRoot>
        <span>
          <text __id="text1">te[st</text>
          <text>goodbye</text>
        </span>
        <span>
          <text>im gonna dissapear</text>
        </span>
        <div>
          <div>
            <text __id="text2">te]st</text>
            <text>moving</text>
            <text>up</text>
          </div>
          <text>staying here</text>
        </div>
      </modelRoot>
```
to this
```
      <modelRoot>
        <span>
          <text __id="text1">te[]st</text>
          <text>moving</text>
          <text>up</text>
        </span>
        <text>staying here</text>
      </modelRoot>
```
currently if the result of the delition are 2 text nodes next to the cursor it will merge them, however this should be easily extendable to other cases.

### CODE REVIEW

**Comment by nvdk** [comment]
just wondering @Asergey91 did you have a look at the 'delete-selection' command? is there a specific reason not to use that here?

**Comment by Asergey91** [comment]
> just wondering @Asergey91 did you have a look at the 'delete-selection' command? is there a specific reason not to use that here? See https://github.com/lblod/ember-rdfa-editor/blob/development/addon/commands/selection-command.ts#L33

yes, however that uses insert/old remove algos under the hood and they dont really do node merges nicely which is needed for backspace/delete hendlers

File: addon/model/operations/remove-operation.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  get nodes(): ModelNode[] {
+    return this._nodes;
+  }
+
+  set nodes(value: ModelNode[]) {
+    this._nodes = value;
+  }
+
+  execute(): void {
+
```
Comment (abeforgit) [inline addon/model/operations/remove-operation.ts:33]: why is this empty?

File: addon/model/operations/operation-algorithms.ts [hunk 2]
```diff
@@ -7,13 +7,103 @@ import RangeMapper, {
...
+      node.remove();
+    });
+
+    const before = range.start.nodeBefore();
+    if (before?.parent && before.index != null) {
+      nodesToMove.forEach((node) => node.remove());
+      before.parent.insertChildrenAtIndex(before.index + 1, ...nodesToMove);
+    }
+
+    //merge logic... needs more work
```
Comment (abeforgit) [inline addon/model/operations/operation-algorithms.ts:93]: what work?

Comment (abeforgit) [inline addon/model/operations/operation-algorithms.ts:93]: what work?

File: addon/model/model-range.ts [hunk 3]
```diff
@@ -259,6 +262,63 @@ export default class ModelRange {
     return this.start.sameAs(this.end);
   }
 
+  normalize(): void {
```
Comment (abeforgit) [inline addon/model/model-range.ts:265]: could you add a doc comment saying what this actually does?

Comment (abeforgit) [inline addon/model/model-range.ts:265]: could you add a doc comment saying what this actually does?

File: addon/model/model-range.ts [hunk 4]
```diff
@@ -259,6 +262,63 @@ export default class ModelRange {
...
+    let flip = false;
+    const length =
+      this.start.path.length > this.end.path.length
+        ? this.start.path.length
+        : this.end.path.length;
+
+    for (let i = 0; i < length; i++) {
+      if (this.start.path[i] == undefined || this.end.path[i] == undefined) {
```
Comment (abeforgit) [inline addon/model/model-range.ts:273]: please use triple equal signs, and explicit type coercions where needed

Comment (abeforgit) [inline addon/model/model-range.ts:273]: please use triple equal signs, and explicit type coercions where needed

File: addon/model/model-range.ts (hunk 3 above)
Comment (abeforgit) [reply addon/model/model-range.ts:265]: Looking at it, it seems like you're making sure that start is before end. On a range, this should always be the case. Where did you encounter this not being so?



Also, modelPosition.compare exists, most of this function just reimplements the same logic

File: addon/model/model-range.ts [hunk 5]
```diff
@@ -259,6 +262,63 @@ export default class ModelRange {
...
+    }
+
+    if (flip) {
+      const buffer = this.start;
+      this.start = this.end;
+      this.end = buffer;
+    }
+  }
+  /**
+   * janky debug function
```
Comment (abeforgit) [inline addon/model/model-range.ts:291]: This seems useful, please un-jank this

Comment (abeforgit) [inline addon/model/model-range.ts:291]: This seems useful, please un-jank this

File: addon/model/model-range.ts [hunk 6]
```diff
@@ -259,6 +262,63 @@ export default class ModelRange {
...
+    }
+    root = root.clone();
+    const range = ModelRange.fromPaths(root, this.start.path, this.end.path);
+
+    const startRange = new ModelRange(range.start, range.start);
+    const endRange = new ModelRange(range.end, range.end);
+    const startText = new ModelText('[===START===]');
+    const endText = new ModelText('[===END===]');
+
+    OperationAlgorithms.insert(endRange, endText);
```
Comment (abeforgit) [inline addon/model/model-range.ts:306]: This breaks a big assumption of operations being the smallest unit of state changes, please use at least an operation. You can pass undefined for the event-bus parameter. (you're working on a copy, so technically it's fine to use the algorithm directly, but it sets a bad precedent.

Comment (abeforgit) [inline addon/model/model-range.ts:306]: This breaks a big assumption of operations being the smallest unit of state changes, please use at least an operation. You can pass undefined for the event-bus parameter. (you're working on a copy, so technically it's fine to use the algorithm directly, but it sets a bad precedent.)

File: addon/model/operations/operation-algorithms.ts [hunk 7]
```diff
@@ -7,13 +7,103 @@ import RangeMapper, {
...
+    const after = range.start.nodeAfter();
+    if (before && after) {
+      if (before.modelNodeType === 'TEXT' && after.modelNodeType === 'TEXT') {
+        (<ModelText>before).content += (<ModelText>after).content;
```
Comment (abeforgit) [inline addon/model/operations/operation-algorithms.ts:97]: please use `thing as Type` syntax for typecasts

File: addon/model/operations/operation-algorithms.ts [hunk 8]
```diff
@@ -7,13 +7,103 @@ import RangeMapper, {
...
+    const before = range.start.nodeBefore();
+    if (before?.parent && before.index != null) {
+      nodesToMove.forEach((node) => node.remove());
+      before.parent.insertChildrenAtIndex(before.index + 1, ...nodesToMove);
+    }
+
+    //merge logic... needs more work
+    const after = range.start.nodeAfter();
+    if (before && after) {
+      if (before.modelNodeType === 'TEXT' && after.modelNodeType === 'TEXT') {
```
Comment (abeforgit) [inline addon/model/operations/operation-algorithms.ts:96]: If you use `ModelNode.isModelText()` you wouldn't need to cast on the next line

Comment (abeforgit) [inline addon/model/operations/operation-algorithms.ts:96]: If you use `ModelNode.isModelText()` you wouldn't need to cast on the next line

File: addon/model/operations/operation-algorithms.ts [hunk 9]
```diff
@@ -7,13 +7,103 @@ import RangeMapper, {
...
+    const openingTagNodes = allNodes.filter((node) => {
+      if (confinedNodes.includes(node)) {
+        return false;
+      } else {
+        return true;
+      }
+    });
+
+    openingTagNodes.forEach((opNode) => {
+      const nodesToUnindent = (<ModelElement>opNode).children;
```
Comment (abeforgit) [inline addon/model/operations/operation-algorithms.ts:75]: please use `thing as Type` syntax for typecasts

Comment (abeforgit) [inline addon/model/operations/operation-algorithms.ts:75]: please use `thing as Type` syntax for typecasts

**Review by abeforgit — CHANGES_REQUESTED** [review]

File: addon/model/model-range.ts (hunk 3 above)
Comment (Asergey91) [reply addon/model/model-range.ts:265]: yes I actaully wanted to ask you abou this, in tests this was the case, range end and range start were always based on the order of parameters passed to the constrcutor.

File: addon/model/operations/operation-algorithms.ts [hunk 10]
```diff
@@ -7,13 +7,182 @@ import RangeMapper, {
...
+    const allNodes: ModelNode[] = [];
+    const walker = new ModelTreeWalker({ range: range });
+    for (const node of walker) {
+      allNodes.push(node);
+    }
+
+    //get all nodes that are fully contained in the range
+    //ie [<span><text>abc</text>]</span>
+    //would grab just the text node
+    const confinedNodes: ModelNode[] = [];
```
Comment (abeforgit) [inline addon/model/operations/operation-algorithms.ts:73]: see prev comment about unpacking iterables

**Review by abeforgit — CHANGES_REQUESTED** [review]

File: addon/model/operations/operation-algorithms.ts [hunk 11]
```diff
@@ -7,13 +7,182 @@ import RangeMapper, {
...
+      if (!cantRemove) {
+        cantRemove =
+          ModelNode.isModelElement(opNode) &&
+          !opNode.getRdfaAttributes().isEmpty;
+      }
+
+      if (cantRemove) {
+        cantRemoveOpeningTagNodes.push(opNode);
+      } else {
+        const nodesToUnindent = (opNode as ModelElement).children;
```
[the remark this answers is not in the document]
Comment (Asergey91) [reply addon/model/operations/operation-algorithms.ts:120]: opening tag node should always be an element... unless we can nest text nodes in each other?

[the remark this answers is not in the document]
Comment (abeforgit) [reply addon/model/operations/operation-algorithms.ts:120]: theres nothing in the code that guarantees that, you're getting all nodes in a range, and then checking if they're also part of the confined nodes (which is a weird check on its own, not sure what that achieves)

File: addon/model/operations/operation-algorithms.ts (hunk 10 above)
Comment (Asergey91) [reply addon/model/operations/operation-algorithms.ts:73]: not sure that i can simplify this

File: addon/model/operations/operation-algorithms.ts [hunk 12]
```diff
@@ -7,13 +7,182 @@ import RangeMapper, {
...
+
+    openingTagNodes.forEach((opNode) => {
+      //check if we can remove it
+      let cantRemove;
+
+      //check in the config const
+      cantRemove =
+        ModelNode.isModelElement(opNode) &&
+        cantRemoveOpeningTagNodeTypes.find((type) => type == opNode.type)
+          ? true
```
[the remark this answers is not in the document]
Comment (Asergey91) [reply addon/model/operations/operation-algorithms.ts:107]: with .has this should no longer be ambiguous right?

**Pull request merged**